src/block-data-providers/bitcoin-core/provider.ts

```diff
@@ -0,0 +1,237 @@
+                currentBlockHeight: 0,
+                indexedBlockHeight:
+                    this.configService.get<BitcoinNetwork>('app.network') ===
+                    BitcoinNetwork.MAINNET
+                        ? TAPROOT_ACTIVATION_HEIGHT - 1
+                        : 0,
+            };
+            await this.setState(state);
+        }
+    }
+
+    @Cron(CronExpression.EVERY_10_SECONDS)
+    async sync() {
+        if (this.isSyncing) return;
+        this.isSyncing = true;
+
+        const state = await this.getState();
+        if (!state) {
+            throw new Error('State not found');
+        }
+        const tipHeight = await this.getTipHeight();
+        if (tipHeight <= state.indexedBlockHeight) {
+            this.logger.debug(
+                `No new blocks found. Current tip height: ${tipHeight}`,
+            );
+            this.isSyncing = false;
+            return;
+        }
+
+        let height = state.indexedBlockHeight + 1;
+
+        for (height; height <= tipHeight; height++) {
+            const transactions = await this.processBlock(height);
+            for (const transaction of transactions) {
+                const { txid, vin, vout, blockHeight, blockHash } = transaction;
+                await this.indexTransaction(
+                    txid,
+                    vin,
+                    vout,
+                    blockHeight,
+                    blockHash,
+                );
+            }
+            state.indexedBlockHeight = height;
+            await this.setState(state);
+        }
+        this.isSyncing = false;
+    }
+
+    private async getTipHeight(): Promise<number> {
+        const body = {
+            method: 'getblockcount',
+            params: [],
+        };
+        return this.request(body);
+    }
+
+    private async getBlockHash(height: number): Promise<string> {
+        const body = {
+            method: 'getblockhash',
+            params: [height],
+        };
+        return this.request(body);
+    }
+
+    private async getBlock(hash: string, verbosity: number): Promise<Block> {
+        const body = {
+            method: 'getblock',
+            params: [hash, verbosity],
+        };
+        return this.request(body);
+    }
+
+    private async getRawTransaction(
+        txid: string,
+        isVerbose: boolean,
+    ): Promise<BlockTransaction> {
+        const body = {
+            method: 'getrawtransaction',
+            params: [txid, isVerbose],
+        };
+        return this.request(body);
+    }
+
+    public async processBlock(height: number): Promise<Transaction[]> {
+        const parsedTransactionList: Transaction[] = [];
+        const blockHash = await this.getBlockHash(height);
+        this.logger.debug(
+            `Processing block at height ${height}, hash ${blockHash}`,
+        );
+        const block = await this.getBlock(blockHash, 2);
+        const excludedCoinbaseTxns = block.tx.slice(1);
```

Comment:
slice is bad, we are unnecessarily creating another copy of the transactions.
Just use a simple for loop like `for(let i = 1; i < block.tx.length; i++)`